core/src/main/scala/kafka/log/Log.scala

```diff
@@ -935,6 +935,23 @@ class Log(@volatile var dir: File,
    */
   def numberOfSegments: Int = segments.size
 
+  /**
+   * Closes this log after taking the snapshot of state of this log (epochs, offsets etc).
+   */
+  def closeWithProducerStateSnapshot(): Unit = {
```

Comment:
We have a method called `takeSnapshot` already. Perhaps it would be ok to modify that and call it directly instead of trying to combine the two operations:
```scala
log.flush()
log.takeProducerSnapshot()
log.close()	          
```

Reply:
reverted.